src/middleware/ajv/index.js

```diff
@@ -0,0 +1,103 @@
+const draftSchema = require('ajv/lib/refs/json-schema-draft-04.json');
+const Ajv = require('ajv');
+
+const { formats } = require('./formats');
+
+const createRequestAjv = (openApiSpec, options={}) => {
+  return createAjv(openApiSpec, options);
+}
+
+const createResponseAjv = (openApiSpec, options={}) => {
+  return createAjv(openApiSpec, options, false);
+}
+
+const createAjv = (openApiSpec, options = {}, request=true) => {
+  const ajv = new Ajv({
+    ...options,
+    schemaId: 'auto',
+    allErrors: true,
+    meta: draftSchema,
+    formats: { ...formats, ...options.formats },
+    unknownFormats: options.unknownFormats,
+  });
+  ajv.removeKeyword('propertyNames');
+  ajv.removeKeyword('contains');
+  ajv.removeKeyword('const');
+  
+  if (request) {
+    ajv.removeKeyword('readOnly');
+    ajv.addKeyword('readOnly', {
+      modifying: true,
+      compile: sch => {
+        if (sch) {
+          return function validate(data, path, obj, propName) {
+            const isValid = !(sch === true && data != null);
+            delete obj[propName];
+            validate.errors = [
+              {
+                keyword: 'readOnly',
+                dataPath: path,
+                message: `is read-only`,
+                params: { readOnly: propName },
+              },
+            ];
+            return isValid;
+          };
+        }
+
+        return () => true;
+      },
+    });
+  } else {
+    // response
+    ajv.removeKeyword('writeOnly');
+    ajv.addKeyword('writeOnly', {
+      modifying: true,
+      compile: sch => {
+        if (sch) {
+          return function validate(data, path, obj, propName) {
```

Comment:
I realize there are 2 different params from this to th eone 20 lines up, but is there any way to bring them together?

Is clarity worth more in this case than brevity?

Reply:
Don't think I understand the question. If you're wondering why it's returning a function, the schema needs to be initialized by the first function and then we pass in the data to be validated on the second time. Therefore initialize the validator once and validate multiple times if needed.

Reply:
Fixed.